source/adiabatic_conditions/initial_profile.cc

```diff
@@ -99,21 +99,36 @@ namespace aspect
           // Handle the case that cp is zero (happens in simple Stokes test problems like sol_cx). By setting
           // 1/cp = 0.0 we will have a constant temperature profile with depth.
           const double one_over_cp = (out.specific_heat[0]>0.0) ? 1.0/out.specific_heat[0] : 0.0;
-          const double gravity = this->get_gravity_model().gravity_vector(representative_point).norm();
+          double gravity = this->get_gravity_model().gravity_vector(representative_point).norm();
+
+          // Check whether gravity is pointing up / out or down / in. In the normal case it should
+          // point down / in and therefore gravity should be positive, leading to increasing
+          // adiabatic pressures and temperatures with depth. In some cases it will point up / out
+          // (e.g. for backward advection), in which case the pressures and temperatures should
+          // decrease with depth and therefore gravity has to be negative in the following equations.
+          const Point<dim> point_surf = this->get_geometry_model().representative_point(0);
+          const Point<dim> point_bot = this->get_geometry_model().representative_point(this->get_geometry_model().maximal_depth());
+          if (g * (point_bot - point_surf) < 0)
+            gravity = - gravity;
```

Comment:
I'm a big fan of making variables `const`. Here, you could achieve this by moving the declaration of `gravity` down and initializing it with
```
  const double gravity = (g * (point_bot - point_surf) >= 0 ?
                                       this->get_gravity_model().gravity_vector(representative_point).norm() :
                                       -this->get_gravity_model().gravity_vector(representative_point).norm());
```

I guess there is a potential problem if the gravity vector does not point straight down (in which case one should only use the vertical component) but that is a problem for a later time to solve.